maven-core/src/main/java/org/apache/maven/plugin/DefaultMojosExecutionStrategy.java

```diff
@@ -0,0 +1,48 @@
+package org.apache.maven.plugin;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.lifecycle.LifecycleExecutionException;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * Default mojo execution strategy.
+ * It just iterates over mojoExecutions list and pass all of them to executor
+ */
+@Named
+@Singleton
+public class DefaultMojosExecutionStrategy implements MojosExecutionStrategy
```

Comment:
Is this the class where you want to have a different implementation for the cache.
I assume all mojos use the same executor, so it is possible to inject the Executor via the constructor?

Reply:
> Is this the class where you want to have a different implementation for the cache.
> I assume all mojos use the same executor, so it is possible to inject the Executor via the constructor?

DefaultMojosExecutionStrategy and Executor in our case have different scopes
First singleton (maven session), second one has project scope

If make both of them singleton then it is required to pass a lot of parameters to execute method (PhaseRecorder, ProjectIndex etc...) most of them related to internal implementation details I'm not sure whether it is ok to make them public API

Reply:
In other words each call of 
``` 
public void execute( final MavenSession session,
                         List<MojoExecution> mojoExecutions,
                         final ProjectIndex projectIndex ) throws LifecycleExecutionException
```

creates new Executor with the full context related to particular project
this allows us to not pass details between interfaces on api level